packages/cli-kit/src/private/node/ui/components/SelectInput.tsx

```diff
@@ -0,0 +1,142 @@
+import React, {useState, useEffect, useRef, useCallback} from 'react'
+import {Box, Key, Text, useApp, useInput} from 'ink'
+import {groupBy, isEqual, mapValues} from 'lodash-es'
+
+export interface Props<T> {
+  items: Item<T>[]
+  onSelect: (item: Item<T>) => void
+}
+
+export interface Item<T> {
+  label: string
+  value: T
+  key?: string
+  group?: string
+}
+
+function groupItems<T>(items: Item<T>[]) {
+  let index = 0
+
+  return mapValues(groupBy(items, 'group'), (groupItems) =>
+    groupItems.map((groupItem) => {
+      const item = {...groupItem, key: groupItem.key ?? (index + 1).toString(), index}
+      index += 1
+      return item
+    }),
+  )
+}
+
+export default function SelectInput<T>({items, onSelect}: React.PropsWithChildren<Props<T>>): JSX.Element | null {
+  const [inputStack, setInputStack] = useState<string | null>(null)
+  const [inputTimeout, setInputTimeout] = useState<NodeJS.Timeout | null>(null)
+  const [selectedIndex, setSelectedIndex] = useState(0)
+  const keys = useRef(new Set(items.map((item) => item.key)))
+  const {exit: unmountInk} = useApp()
+  const groupedItems = groupItems(items)
+  const groupTitles = Object.keys(groupedItems)
+
+  const previousItems = useRef<Item<T>[]>(items)
+
+  useEffect(() => {
```

Comment:
Have you considered increasing the readability of this component by extracting the hooks invocations into hooks with descriptive names?

```js
useSelectFirstItemHook(items)
useHandleInputCallback(items, (selectedIndex) => { })
// ...
```

Reply:
Unless we're going to be reusing these hooks somewhere else I prefer to put comments above this if we need to explain what they do. It seems to me it's more in react's style to not create too many custom hooks and instead go for simple ones when possible.